### ISSUE TITLE
Pass down write callable in t.w.template

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8300]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)| @mithrandi reported|
|-|-|
|Trac ID|trac#8300|
|Type|enhancement|
|Created|2016-04-17 07:38:53Z|
|Branch|https://github.com/twisted/twisted/tree/flatten-optimization-8300|

Currently the inner flattening loop in `t.w.template` yields a lot of tiny strings, which has a lot of overhead. However, the outer flattening loop just passes these strings to a `write` callable, so we can optimize this by passing `write` down into the inner flattener and calling it directly. A [similar optimization in Nevow](https://github.com/twisted/nevow/pull/79) yielded a significant performance improvement on PyPy (and a modest one even on CPython).

<details><summary>Searchable metadata</summary>

```
trac-id__8300 8300
type__enhancement enhancement
reporter__mithrandi mithrandi
priority__normal normal
milestone__None None
branch__branches_flatten_optimization_8300 branches/flatten-optimization-8300
branch_author__mithrandi mithrandi
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1460878733645594 1460878733645594
changetime__1461057270710979 1461057270710979
version__None None
owner__mithrandi mithrandi

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_1"></a>|@mithrandi commented|
|-|-|

(In [47280]) Branching to flatten-optimization-8300.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_2"></a>|@mithrandi commented|
|-|-|

(In [47280]) Branching to flatten-optimization-8300.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_3"></a>|@mithrandi removed owner|
|-|-|

Builds are running, and benchmark results are here: https://gist.github.com/mithrandi/54845e061217b98cdf4dd7700277f6a2#file-results-rst

I'm trying to get a benchmark into speedcenter [here](https://github.com/twisted-infra/twisted-benchmarks/pull/1), so maybe we should wait for that before merging this.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @mithrandi|
|-|-|

You left in a couple of `assert False`s, but otherwise, this looks good.

I forced a couple of builds to get py27 and pypy baselines for `web_template`, visible here: http://speed.twistedmatrix.com/timeline/#/?exe=4,3,2,5&base=none&ben=web_template&env=1&revs=50&equid=off

Please kill the `assert` lines, then land.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_5"></a>|@mithrandi set status to `closed`|
|-|-|

(In [47298]) Merge flatten-optimization-8300: Optimize template flattening.

Author: mithrandi
Reviewer: glyph
Fixes: [#8300](https://github.com/twisted/twisted/issues/8300)

Optimize template flattening by passing the write callable down into the
flattener and calling it directly, instead of yielding strings (often
very small strings) back up to the driver.

This yields a significant speed increase on PyPy, and a moderate one on
CPython.
